woocommerce/google-listings-and-ads
# Ensure coupon brand restrictions are uploaded to Google Merchant Center.

### Changes proposed in this Pull Request:



Closes #2693.

Since WooCommerce 9.4, [WooCommerce Brands is now part of the WooCommerce core](https://developer.woocommerce.com/2024/10/01/introducing-brands/), the merchant can add brand restrictions when editing coupons, but G4W plugin does not upload the corresponding coupon brand restrictions to Google Merchant Center.

This PR adds coupon restrictions for brands when generating the coupon (promotions) payload for updating to Google Merchant Center.

Note that `WC_Coupon` class does not have a method to get the brand restrictions [like other restrictions do](https://github.com/woocommerce/woocommerce/blob/1a3ff575caf0a060a24a914bd8e08b94bedd6153/plugins/woocommerce/includes/class-wc-coupon.php#L305-L318). This PR uses [get_post_meta()](https://developer.wordpress.org/reference/functions/get_post_meta/) to get the brands from a coupon, then uses [get_objects_in_term()](https://developer.wordpress.org/reference/functions/get_objects_in_term/) to get a list of product IDs from that brand.

Refer to the following tables to better understand the data:
- `wp_postmeta`
    - Store the brands for coupons, queried by`get_post_meta()`
    - Meta keys: `product_brands` and `exclude_product_brands`
- `wp_terms`
    - Where the brand ID, brand name, and brand slug are stored
- `wp_term_taxonomy`
    - Store the taxonomy for the brand (`product_brand`)
- `wp_term_relationships`
    - Store the mapping of products and brand. Queried by `get_objects_in_term()`
    - Should specify the brand taxonomy `product_brand` in `get_objects_in_term()`

**Updated 28th Nov**

For querying brands from a coupon from post_meta, I found in [class-wc-brands-brand-settings-manager.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-brands-brand-settings-manager.php) I can use it like:

```php
\WC_Brands_Brand_Settings_Manager::set_brand_settings_on_coupon( $coupon );
$brand_settings = \WC_Brands_Brand_Settings_Manager::get_brand_settings_on_coupon( $coupon );
```

However, since we do not expect all merchants update to latest WooCommerce, we also need to ensure backward compatibility for those who still use WC Brands plugin. The code above was available [since WC Brands plugin version 1.6.65](https://github.com/woocommerce/woocommerce-brands/blob/0899036f7a68b9b5f17c4d2fd806e7614e27739a/includes/class-wc-brands-brand-settings-manager.php#L8), so I lean towards to still querying post_meta manually as it's safest for all version and in core.

### Screenshots:


### Detailed test instructions:


1. Build the project using this PR, and ensure WooCommerce >= 9.4 is installed on your site
2. Enable Brands feature by updating two options `wc_feature_woocommerce_brands_enabled` and `woocommerce_remote_variant_assignment`
    ```bash
    wp option update wc_feature_woocommerce_brands_enabled 'yes'
    wp option update woocommerce_remote_variant_assignment 2
    ```
3. Create some brands from `Products > Brands`, e.g.
    <img width="1703" alt="Screenshot 2024-11-27 at 14 58 11" src="https://github.com/user-attachments/assets/dfbfaff4-50e2-4bff-9d3b-a3f3202fe54d">
4. Add some products to each of the brand that just created, e.g.

https://github.com/user-attachments/assets/1a6c0fac-2939-408d-9ade-96ba27ac45d8

5. Create a coupon, add brand restrictions to it
    - Make sure your store country supports coupons, e.g. US, GB.
    - Make sure to choose **Show coupon on Google** in **Channel visibility**
    - Note that the inclusion and exclusion cannot be set at the same time, otherwise Google API will respond 400 bad request. It also applies to existing restrictions like products or product categories.
    - <img width="1541" alt="Screenshot 2024-11-27 at 15 14 06" src="https://github.com/user-attachments/assets/7e282aa8-7423-4d03-a403-2181d81d9f82">
6. Save the coupon, go to `WooCommerce > Status > Scheduled Actions`, search for the job `gla/jobs/update_coupon/process_item`, run it manually if it is still pending to save time
7. Wait for a while, go to Google Merchant Center, click **Promotions** from the left menu.
8. Click on the coupon (promotion) you just uploaded, then click **Next**. Ensure that the filters display the product IDs associated with the brand set as a restriction.

https://github.com/user-attachments/assets/01c893c3-53ef-494f-b73c-e3dcfaacb8c8

9. Play around the existing restrictions, make sure they are still working
10. Disable the Brands in core feature (see step 2), make sure the PR won't break it
11. Using WooCommerce beta tester to downgrade WC to 9.2, and install WC Brands plugin, make sure the PR still works.

#### Product added or removed from a brand

1. Let's say now that the coupon brand restriction is synced to GMC after the above test.
2. Go to the product editor, deselect `Test Brand 1` and update the product
    <img width="334" alt="Screenshot 2024-12-05 at 15 09 44" src="https://github.com/user-attachments/assets/cd7cf70f-21bf-4017-a45c-bd8fb0f723f7">
3. Go to `WooCommerce > Status > Scheduled Actions`, search for the job `gla/jobs/update_coupon/process_item`, run it manually if it is still pending to save time
4. Wait for a while, go to Google Merchant Center, click **Promotions** from the left menu.
5. Click on the coupon (promotion) you just uploaded, then click **Next**. Ensure that the filters no longer include the product that was just excluded from the brand.
6. Go to the product editor, add the brand `Test Brand 1` back. You should see the product gets synced to GMC in the coupon filters again.
7. Create a new product and add the brand `Test Brand 1`, you should see the product gets synced to GMC in the coupon filters again.


### Additional details:


### Changelog entry

> Fix - Ensure coupon brand restrictions are uploaded to Google Merchant Center.
